File: lecture_summarizer/telemetry.py
from __future__ import annotations

import json
from dataclasses import dataclass, field
from datetime import datetime, timezone
from pathlib import Path
from threading import Lock
from typing import Dict, Iterable, List


@dataclass(frozen=True)
class RequestEvent:
    """Single Gemini API interaction captured for monitoring."""

    model: str
    modality: str
    started_at: datetime
    finished_at: datetime
    input_tokens: int | None
    output_tokens: int | None
    total_tokens: int | None
    metadata: Dict[str, object] = field(default_factory=dict)

    @property
    def duration_seconds(self) -> float:
        return max((self.finished_at - self.started_at).total_seconds(), 0.0)

    def to_dict(self) -> Dict[str, object]:
        return {
            "model": self.model,
            "modality": self.modality,
            "started_at": self.started_at.isoformat(),
            "finished_at": self.finished_at.isoformat(),
            "duration_seconds": self.duration_seconds,
            "input_tokens": self.input_tokens,
            "output_tokens": self.output_tokens,
            "total_tokens": self.total_tokens,
            "metadata": dict(self.metadata),
        }


@dataclass(frozen=True)
class RunSummary:
    total_requests: int
    total_input_tokens: int
    total_output_tokens: int
    total_tokens: int
    total_duration_seconds: float
    by_model: Dict[str, Dict[str, float]]
    by_modality: Dict[str, Dict[str, float]]

    def to_dict(self) -> Dict[str, object]:
        return {
            "total_requests": self.total_requests,
            "total_input_tokens": self.total_input_tokens,
            "total_output_tokens": self.total_output_tokens,
            "total_tokens": self.total_tokens,
            "total_duration_seconds": self.total_duration_seconds,
            "by_model": self.by_model,
            "by_modality": self.by_modality,
        }


class RunMonitor:
    """Collects Gemini request telemetry for the lifetime of a pipeline run."""

    def __init__(self) -> None:
        self._lock = Lock()
        self._events: List[RequestEvent] = []
        self._notes: List[Dict[str, object]] = []

    def record(self, event: RequestEvent) -> None:
        with self._lock:
            self._events.append(event)

    def note_event(self, name: str, payload: Dict[str, object] | None = None) -> None:
        with self._lock:
            self._notes.append(
                {
                    "name": name,
                    "payload": dict(payload or {}),
                    "timestamp": datetime.now(timezone.utc).isoformat(),
                }
            )

    def events(self) -> List[RequestEvent]:
        with self._lock:
            return list(self._events)

    def notes(self) -> List[Dict[str, object]]:
        with self._lock:
            return list(self._notes)

    def summarize(self) -> RunSummary:
        events = self.events()
        total_requests = len(events)
        total_input_tokens = sum(e.input_tokens or 0 for e in events)
        total_output_tokens = sum(e.output_tokens or 0 for e in events)
        total_tokens = sum(e.total_tokens or 0 for e in events)
        total_duration = sum(e.duration_seconds for e in events)

        by_model: Dict[str, Dict[str, float]] = {}
        by_modality: Dict[str, Dict[str, float]] = {}

        def _update(bucket: Dict[str, Dict[str, float]], key: str, event: RequestEvent) -> None:
            stats = bucket.setdefault(
                key,
                {
                    "requests": 0,
                    "input_tokens": 0,
                    "output_tokens": 0,
                    "total_tokens": 0,
                    "total_duration_seconds": 0.0,
                },
            )
            stats["requests"] += 1
            stats["input_tokens"] += event.input_tokens or 0
            stats["output_tokens"] += event.output_tokens or 0
            stats["total_tokens"] += event.total_tokens or 0
            stats["total_duration_seconds"] += event.duration_seconds

        for ev in events:
            _update(by_model, ev.model, ev)
            _update(by_modality, ev.modality, ev)

        return RunSummary(
            total_requests=total_requests,
            total_input_tokens=total_input_tokens,
            total_output_tokens=total_output_tokens,
            total_tokens=total_tokens,
            total_duration_seconds=total_duration,
            by_model=by_model,
            by_modality=by_modality,
        )

    def costs(self) -> "CostSummary":
        from .costs import estimate_costs

        return estimate_costs(self.events())

    def flush_summary(self, *, to: Path, cost: "CostEstimator") -> None:
        summary = self.summarize()
        costs = cost.estimate(self.events())
        payload = {
            "summary": summary.to_dict(),
            "costs": costs.to_dict(),
            "notes": self.notes(),
        }
        to.parent.mkdir(parents=True, exist_ok=True)
        to.write_text(json.dumps(payload, indent=2, sort_keys=True))

File: src/telemetry.rs
use crate::core::Job;
use crate::cost::CostEstimator;
use crate::utils::ensure_dir;
use serde::Serialize;
use serde_json::json;
use std::collections::HashMap;
use std::fs::File;
use std::io::Write;
use std::path::{Path, PathBuf};
use std::sync::{Arc, Mutex};
use time::format_description::well_known::Rfc3339;
use time::{Duration, OffsetDateTime};

#[derive(Debug, Clone, Serialize)]
pub struct RequestEvent {
    pub model: String,
    pub modality: String,
    #[serde(with = "time::serde::rfc3339")]
    pub started_at: OffsetDateTime,
    #[serde(with = "time::serde::rfc3339")]
    pub finished_at: OffsetDateTime,
    pub input_tokens: Option<u32>,
    pub output_tokens: Option<u32>,
    pub total_tokens: Option<u32>,
    #[serde(default)]
    pub metadata: HashMap<String, serde_json::Value>,
}

impl RequestEvent {
    pub fn duration_seconds(&self) -> f64 {
        (self.finished_at - self.started_at)
            .as_seconds_f64()
            .max(0.0)
    }
}

#[derive(Debug, Default, Serialize)]
pub struct RunSummary {
    pub total_requests: usize,
    pub total_input_tokens: u64,
    pub total_output_tokens: u64,
    pub total_tokens: u64,
    pub total_duration_seconds: f64,
    pub by_model: HashMap<String, SummaryBucket>,
    pub by_modality: HashMap<String, SummaryBucket>,
}

#[derive(Debug, Default, Serialize)]
pub struct SummaryBucket {
    pub requests: usize,
    pub input_tokens: u64,
    pub output_tokens: u64,
    pub total_tokens: u64,
    pub total_duration_seconds: f64,
}

#[derive(Clone)]
pub struct RunMonitor {
    inner: Arc<Mutex<RunState>>,
}

impl Default for RunMonitor {
    fn default() -> Self {
        Self {
            inner: Arc::new(Mutex::new(RunState::default())),
        }
    }
}

#[derive(Default)]
struct RunState {
    events: Vec<RequestEvent>,
    notes: Vec<Note>,
    first_started: Option<OffsetDateTime>,
    last_finished: Option<OffsetDateTime>,
}

#[derive(Debug, Clone, Serialize)]
struct Note {
    name: String,
    payload: serde_json::Value,
    #[serde(with = "time::serde::rfc3339")]
    timestamp: OffsetDateTime,
}

impl RunMonitor {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn record(&self, event: RequestEvent) {
        let mut state = self.inner.lock().unwrap();
        if state.first_started.is_none()
            || event.started_at < state.first_started.unwrap_or(event.started_at)
        {
            state.first_started = Some(event.started_at);
        }
        if state.last_finished.is_none()
            || event.finished_at > state.last_finished.unwrap_or(event.finished_at)
        {
            state.last_finished = Some(event.finished_at);
        }
        state.events.push(event);
    }

    pub fn note_event(&self, name: &str, payload: serde_json::Value) {
        let mut state = self.inner.lock().unwrap();
        state.notes.push(Note {
            name: name.to_string(),
            payload,
            timestamp: OffsetDateTime::now_utc(),
        });
    }

    pub fn events(&self) -> Vec<RequestEvent> {
        self.inner.lock().unwrap().events.clone()
    }

    pub fn summarize(&self) -> RunSummary {
        let state = self.inner.lock().unwrap();
        let mut summary = RunSummary::default();
        summary.total_requests = state.events.len();
        for event in &state.events {
            let input = event.input_tokens.unwrap_or_else(|| {
                event
                    .total_tokens
                    .and_then(|total| event.output_tokens.map(|out| total.saturating_sub(out)))
                    .unwrap_or(0)
            }) as u64;
            let output = event
                .output_tokens
                .or_else(|| {
                    event
                        .total_tokens
                        .and_then(|total| event.input_tokens.map(|inp| total.saturating_sub(inp)))
                })
                .unwrap_or(event.total_tokens.unwrap_or(0)) as u64;
            let total = event.total_tokens.unwrap_or((input + output) as u32) as u64;

            summary.total_input_tokens += input;
            summary.total_output_tokens += output;
            summary.total_tokens += total;
            summary.total_duration_seconds += event.duration_seconds();

            update_bucket(
                summary.by_model.entry(event.model.clone()).or_default(),
                input,
                output,
                total,
                event.duration_seconds(),
            );
            update_bucket(
                summary
                    .by_modality
                    .entry(event.modality.clone())
                    .or_default(),
                input,
                output,
                total,
                event.duration_seconds(),
            );
        }
        summary
    }

    pub fn flush_summary(
        &self,
        to: &Path,
        cost: &CostEstimator,
        job: &Job,
        files: &[PathBuf],
        limits: &HashMap<&str, Option<u32>>,
        ndjson: Option<&Path>,
    ) -> anyhow::Result<()> {
        if let Some(parent) = to.parent() {
            ensure_dir(parent)?;
        }
        let summary = self.summarize();
        let events = self.events();
        let costs = cost.estimate(&events);
        let state = self.inner.lock().unwrap();
        let start = state.first_started.map(|t| t.format(&Rfc3339).unwrap());
        let end = state.last_finished.map(|t| t.format(&Rfc3339).unwrap());
        let elapsed = match (state.first_started, state.last_finished) {
            (Some(s), Some(f)) => (f - s).max(Duration::ZERO).as_seconds_f64(),
            _ => 0.0,
        };

        let payload = json!({
            "job": {
                "source": job.source,
                "kind": job.kind.map(|k| k.as_str().to_string()),
                "model": job.model,
            },
            "totals": {
                "requests": summary.total_requests,
                "input_tokens": summary.total_input_tokens,
                "output_tokens": summary.total_output_tokens,
                "est_cost_usd": (costs.total_cost * 1_000_000.0).round() / 1_000_000.0,
            },
            "time": {
                "start": start,
                "end": end,
                "elapsed_sec": elapsed,
            },
            "limits": limits.iter().map(|(k, v)| (k.to_string(), v)).collect::<HashMap<_, _>>(),
            "files": files.iter().map(|p| p.to_string_lossy().to_string()).collect::<Vec<_>>(),
            "warnings": if costs.estimated { vec!["costs include estimates".to_string()] } else { Vec::new() },
            "notes": state.notes.clone(),
        });

        let mut file = File::create(to)?;
        file.write_all(serde_json::to_string_pretty(&payload)?.as_bytes())?;

        if let Some(ndjson_path) = ndjson {
            if let Some(parent) = ndjson_path.parent() {
                ensure_dir(parent)?;
            }
            let mut ndjson_file = File::create(ndjson_path)?;
            for event in events {
                let line = json!({
                    "model": event.model,
                    "modality": event.modality,
                    "chunk_index": event.metadata.get("chunk_index"),
                    "start_utc": event.started_at.format(&Rfc3339).ok(),
                    "end_utc": event.finished_at.format(&Rfc3339).ok(),
                    "latency_ms": (event.duration_seconds() * 1000.0).round() as i64,
                    "tokens_in": event.input_tokens,
                    "tokens_out": event.output_tokens,
                    "video_start": event.metadata.get("chunk_start_seconds"),
                    "video_end": event.metadata.get("chunk_end_seconds"),
                    "file_uri": event.metadata.get("file_uri"),
                    "manifest_path": event.metadata.get("manifest_path"),
                    "response_path": event.metadata.get("response_path"),
                });
                ndjson_file.write_all(serde_json::to_string(&line)?.as_bytes())?;
                ndjson_file.write_all(b"\n")?;
            }
        }
        Ok(())
    }
}

fn update_bucket(bucket: &mut SummaryBucket, input: u64, output: u64, total: u64, duration: f64) {
    bucket.requests += 1;
    bucket.input_tokens += input;
    bucket.output_tokens += output;
    bucket.total_tokens += total;
    bucket.total_duration_seconds += duration;
}
